text/logs/0178-mapping-to-otlp-anyvalue.md

```diff
@@ -0,0 +1,217 @@
+field.
+
+If the keys of the source array are not strings they MUST be converted to
+strings by any means available, often via a toString() or stringify functions
+available in programming languages. The conversion function MUST be chosen in a
+way that ensures that the resulting string keys are unique in the target array.
+
+The value part of each element of the source array SHOULD be converted to
+AnyValue recursively.
+
+For example a JSON object `{"a": 123, "b": "def"}` SHOULD be converted to
+
+```
+AnyValue{
+    kvlist_value:KeyValueList{
+        values:[
+            KeyValue{key:"a",value:AnyValue{int_value:123}},
+            KeyValue{key:"b",value:AnyValue{string_value:"def"}},
+        ]
+    }
+}
+```
+
+#### Associative Arrays With Non-Unique Keys
+
+Values that represent an associative arrays with non-unique keys where multiple values may be associated with the same key (also sometimes known
+as multimaps, multidicts) SHOULD be converted to AnyValue's
+[kvlist_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L36)
+field.
+
+The resulting
+[kvlist_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L36)
+field MUST list each key only once and the value of each element of
+[kvlist_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L36)
+field MUST be an array represented using AnyValue's
+[array_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L35)
+field, each element of the
+[array_value](https://github.com/open-telemetry/opentelemetry-proto/blob/38b5b9b6e5257c6500a843f7fdacf89dd95833e8/opentelemetry/proto/common/v1/common.proto#L35)
+representing one value of source array associated with the given key.
+
+For example an associative array shown in the following table:
+
+|Key|Value|
+|---|---|
+|"abc"|123|
+|"def"|"foo"|
+|"def"|"bar"|
+
+SHOULD be converted to:
+
+```
+AnyValue{
+    kvlist_value:KeyValueList{
+        values:[
+            KeyValue{
+                key:"abc",
+                value:AnyValue{array_value:ArrayValue{values[
+                    AnyValue{int_value:123}
+                ]}}
+            },
+            KeyValue{
+                key:"def",
+                value:AnyValue{array_value:ArrayValue{values[
+                    AnyValue{string_value:"foo"},
+                    AnyValue{string_value:"bar"}
+                ]}}
+            },
+        ]
+    }
+}
+```
+
+### Other Values
+
+Any other values not listed above SHOULD be converted to AnyValue's
```

Comment:
Should we define an expected mapping for a set? 

[Many languages](https://en.wikipedia.org/wiki/Set_(abstract_data_type)#Language_support) have a representation, so it seems we can either establish an imperfect but consistent mapping, or expect to have a variety of inconsistent mappings.

Perhaps the best representation would be as a `kvlist_value`, where each `key` represents an item in the set and each `value` is `empty`.

Reply:
> Should we define an expected mapping for a set?
> 
> [Many languages](https://en.wikipedia.org/wiki/Set_(abstract_data_type)#Language_support) have a representation, so it seems we can either establish an imperfect but consistent mapping, or expect to have a variety of inconsistent mappings.

That's a good point

> Perhaps the best representation would be as a `kvlist_value`, where each `key` represents an item in the set and each `value` is `empty`.

This would limit the `key` type to `string` though, wouldn't it? Perhaps we could simply recommend using Arrays for sets?

Reply:
> This would limit the type though as key must be string

Yeah, that's true. 

> recommend using Arrays for sets

I agree with this.

Reply:
Added a section for Sets.